core/engine/engine.go

```diff
@@ -1,99 +1,166 @@
 package engine
 
 import (
-	"bytes"
 	"context"
-	"io"
 
 	"github.com/filecoin-project/go-indexer-core"
+	"github.com/filecoin-project/indexer-reference-provider/config"
 	"github.com/filecoin-project/indexer-reference-provider/core"
+	schema "github.com/filecoin-project/storetheindex/api/v0/ingest/schema"
 	"github.com/ipfs/go-cid"
 	"github.com/ipfs/go-datastore"
 	logging "github.com/ipfs/go-log/v2"
 	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	crypto "github.com/libp2p/go-libp2p-core/crypto"
 	"github.com/libp2p/go-libp2p-core/host"
 	peer "github.com/libp2p/go-libp2p-core/peer"
 	legs "github.com/willscott/go-legs"
-
-	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
 )
 
 var log = logging.Logger("reference-provider")
 
 var _ core.Interface = &Engine{}
 
+const (
+	// NOTE: For now this is an IPLD-aware reference provider, wo
+	// we need to force GraphSupport. This may change in the future
+	// and we may pass this as an option.
+	graphSupport = true
+)
+
 // Engine is an implementation of the core reference provider interface.
 type Engine struct {
+	// Provider's privateKey
+	privKey crypto.PrivKey
 	// Host running the provider process.
 	host host.Host
+	// Linksystem used for reference provider.
+	lsys ipld.LinkSystem
 	// Datastore used for persistence of different assets
 	// (advertisements, indexed data, etc.).
 	ds datastore.Batching
-	// Local index with all provided data.
-	// NOTE: We may want to wrap indexer.Interface into datastore interface
-	// to be able to use it as a datastore right away.
-	index indexer.Interface
-	// go-legs subscriber for IPLD-aware ingestion
-	ipldIngest legs.LegPublisher
+	lp legs.LegPublisher
 	// pubsubtopic where the provider will push advertisements
 	pubSubTopic string
+
+	// Local indexer with all provided data.
+	// NOTE: It may be the case that providers want to keep a local index with all
+	// provider data, and they want to store it with go-indexer-core. In this case
+	// we may need to wrap go-indexer-core so linkSystems can use it.
+	//indexer indexer.Interface
```

Comment:
will we be able to efficiently access individual updates that have been indexed into go-indexer-core? I was expecting that would be lossy (or at least inefficient to extract back out the reverse index), and we'd want/need to have the cids split into individual updates stored separately.

Reply:
This was actually the placeholder to ask you this exact question. I left this in case we want to support individual indexing from the reference provider, if this is not the case this can be removed. We only keep advertisements and updates in the datastore.

Reply:
i think we don't need to support individual indexing for now